plonky2/src/recursion/cyclic_recursion.rs

```diff
@@ -0,0 +1,490 @@
+    pub dummy_verifier_data: VerifierCircuitTarget,
+    pub base_case: BoolTarget,
+}
+
+pub struct CyclicPublicInputs<
+    F: RichField + Extendable<D>,
+    C: GenericConfig<D, F = F>,
+    const D: usize,
+> {
+    pub circuit_digest: HashOut<F>,
+    pub constants_sigmas_cap: MerkleCap<F, C::Hasher>,
+    pub base_case: bool,
+}
+
+impl<F: RichField + Extendable<D>, C: GenericConfig<D, F = F>, const D: usize>
+    CyclicPublicInputs<F, C, D>
+{
+    fn from_slice(slice: &[F], common_data: &CommonCircuitData<F, C, D>) -> Result<Self>
+    where
+        C::Hasher: AlgebraicHasher<F>,
+    {
+        // The structure of the public inputs is `[...,circuit_digest, constants_sigmas_cap, base_case]`.
+        let cap_len = common_data.config.fri_config.num_cap_elements();
+        let len = slice.len();
+        ensure!(len >= 4 + 4 * cap_len + 1, "Not enough public inputs");
+        let base_case = slice[len - 1];
+        ensure!(
+            base_case.is_one() || base_case.is_zero(),
+            "Base case flag {:?} is not binary",
+            base_case
+        );
+        let constants_sigmas_cap = MerkleCap(
+            (0..cap_len)
+                .map(|i| HashOut {
+                    elements: std::array::from_fn(|j| slice[len - 1 - 4 * (cap_len - i) + j]),
+                })
+                .collect(),
+        );
+        let circuit_digest =
+            HashOut::from_partial(&slice[len - 5 - 4 * cap_len..len - 1 - 4 * cap_len]);
+
+        Ok(Self {
+            circuit_digest,
+            constants_sigmas_cap,
+            base_case: base_case.is_one(),
+        })
+    }
+}
+
+pub struct CyclicPublicInputsTarget {
+    pub circuit_digest: HashOutTarget,
+    pub constants_sigmas_cap: MerkleCapTarget,
+    pub base_case: BoolTarget,
+}
+
+impl CyclicPublicInputsTarget {
+    fn from_slice<F: RichField + Extendable<D>, C: GenericConfig<D, F = F>, const D: usize>(
+        slice: &[Target],
+        common_data: &CommonCircuitData<F, C, D>,
+    ) -> Result<Self> {
+        let cap_len = common_data.config.fri_config.num_cap_elements();
+        let len = slice.len();
+        ensure!(len >= 4 + 4 * cap_len + 1, "Not enough public inputs");
+        let base_case = BoolTarget::new_unsafe(slice[len - 1]);
+        let constants_sigmas_cap = MerkleCapTarget(
+            (0..cap_len)
+                .map(|i| HashOutTarget {
+                    elements: std::array::from_fn(|j| slice[len - 1 - 4 * (cap_len - i) + j]),
+                })
+                .collect(),
+        );
+        let circuit_digest = HashOutTarget {
+            elements: std::array::from_fn(|i| slice[len - 5 - 4 * cap_len + i]),
+        };
+
+        Ok(Self {
+            circuit_digest,
+            constants_sigmas_cap,
+            base_case,
+        })
+    }
+}
+
+impl<F: RichField + Extendable<D>, const D: usize> CircuitBuilder<F, D> {
+    pub fn cyclic_recursion<C: GenericConfig<D, F = F>>(
+        mut self,
```

Comment:
In our aggregation circuit we'll want two cyclic recursion gadgets; can we have this take `&mut self` so we can call it twice?

My understanding is this currently consumes `self` so that it can do the `data.common == common_data` check after `build()`. Perhaps we could instead have `CircuitBuilder` hold an `Option<CommonCircuitData<...>>`, tracking any `common_data` passed to `cyclic_recursion`, and then have `build()` check that it matches.

Alternatively we could continue having `cyclic_recursion` consume `self`, but generalize it so it can add multiple instances of a cyclic recursion gadget. Might be a more complicated API though.

Reply:
> In our aggregation circuit we'll want two cyclic recursion gadgets; can we have this take `&mut self` so we can call it twice?
> 
> My understanding is this currently consumes `self` so that it can do the `data.common == common_data` check after `build()`. Perhaps we could instead have `CircuitBuilder` hold an `Option<CommonCircuitData<...>>`, tracking any `common_data` passed to `cyclic_recursion`, and then have `build()` check that it matches.

That sounds good to me! Do you think we should do that on the `CircuitBuilder` or add a wrapped version `CyclicCircuitBuilder` and add it there (to avoid cyclic recursion logic in the base circuit builder)?

Reply:
I think having it in `CircuitBuilder` seems okay to me

Reply:
Added `goal_common_data: Option<CommonCircuitData<F, D>>` to `CircuitBuiler`.